mempool/mempool_test.go

```diff
@@ -166,10 +168,84 @@ func (s *fakeChain) SetPastMedianTime(medianTime time.Time) {
 // CalcSequenceLock returns the current sequence lock for the passed transaction
 // associated with the fake chain instance.
 func (s *fakeChain) CalcSequenceLock(tx *dcrutil.Tx, view *blockchain.UtxoViewpoint) (*blockchain.SequenceLock, error) {
-	return &blockchain.SequenceLock{
-		MinHeight: -1,
-		MinTime:   -1,
-	}, nil
+	// A value of -1 for each lock type allows a transaction to be included in a
+	// block at any given height or time.
+	sequenceLock := &blockchain.SequenceLock{MinHeight: -1, MinTime: -1}
+
+	// Sequence locks do not apply if the tx version is less than 2, or the tx
+	// is a coinbase or stakebase, so return now with a sequence lock that
+	// indicates the tx can possibly be included in a block at any given height
+	// or time.
+	msgTx := tx.MsgTx()
+	enforce := msgTx.Version >= 2
+	if !enforce || blockchain.IsCoinBaseTx(msgTx) || stake.IsSSGen(msgTx) {
+		return sequenceLock, nil
+	}
+
+	for txInIndex, txIn := range msgTx.TxIn {
+		// Nothing to calculate for this input when relative time locks are
+		// disabled for it.
+		sequenceNum := txIn.Sequence
+		if sequenceNum&wire.SequenceLockTimeDisabled != 0 {
+			continue
+		}
+
+		utxo := view.LookupEntry(&txIn.PreviousOutPoint.Hash)
+		if utxo == nil {
+			str := fmt.Sprintf("output %v referenced from transaction %s:%d "+
+				"either does not exist or has already been spent",
+				txIn.PreviousOutPoint, tx.Hash(), txInIndex)
+			return nil, blockchain.RuleError{
+				ErrorCode:   blockchain.ErrMissingTxOut,
+				Description: str,
+			}
+		}
+
+		// Calculate the sequence locks from the point of view of the next block
+		// for inputs that are in the mempool.
+		inputHeight := utxo.BlockHeight()
+		if inputHeight == mining.UnminedHeight {
+			inputHeight = s.BestHeight() + 1
+		}
+
+		// Mask off the value portion of the sequence number to obtain
+		// the time lock delta required before this input can be spent.
+		// The relative lock can be time based or block based.
+		relativeLock := int64(sequenceNum & wire.SequenceLockTimeMask)
+
+		if sequenceNum&wire.SequenceLockTimeIsSeconds != 0 {
```

Comment:
could be
```go
 if relativeLock != 0 {
 ```

Reply:
They aren't the same thing.  The relative lock is the sequence number masked with 0xfff, or, in other words the bottom 16 bits.  On the other hand, the bit which if the relative lock is to be interpreted as
either a relative amount of seconds or a relative block height is bit 22 of the sequence number.

Reply:
oh, i misread the variable name.